File: WebChat.Client/src/types/Conversation.ts
import { ConversationType } from "../enums/ConversationType";
import { MessageStatus } from "../enums/MessageStatus";
import FileMetadata from "./FileMetadata";

export interface ConversationMember {
  conversationId: string;
  userId: string;
  isHidden: boolean;
  receivedToId: number;
  seenToId: number;
  loadFromId: number;
  hiddenMessageIds: Array<number>;
  isBlock: boolean;
}

export interface Message {
  isDeleted: boolean;
  messageId: number;
  conversationId: string;
  fromUserId: string;
  sendTime: Date;
  text: string;
  status: MessageStatus;
  files: Array<FileMetadata>;
}

export default interface Conversation {
  type: ConversationType;
  conversationId: string;
  isHidden: boolean;
  receivedToId: number;
  seenToId: number;
  loadFromId: number;
  messages: Array<Message>;
  hiddenMessageIds: Array<number>;
  isBlock: boolean;
  status: MessageStatus;
}

File: WebChat.Domain/Entities/Conversation.cs
namespace WebChat.Domain.Entities;

public class Conversation : AuditableEntity, IAggregateRoot
{
    public ConversationType Type { get; set; } = ConversationType.Unknown;

    public Guid ConversationId { get; set; } = Guid.Empty;

    public virtual ICollection<ConversationMember> Members { get; set; } = [];

    public virtual ICollection<Message> Messages { get; set; } = [];

    public override Guid GetId() => ConversationId;
}

public class ConversationMember : AuditableEntity<(Guid ConversationId, Guid UserId)>
{
    public Guid ConversationId { get; set; } = Guid.Empty;

    public Guid UserId { get; set; } = Guid.Empty;

    public bool IsHidden { get; set; } = false;

    public long ReceivedToId { get; set; } = 0;

    public long SeenToId { get; set; } = 0;

    public long LoadFromId { get; set; } = 0;


    public IList<long> HiddenMessageIds { get; set; } = [];

    public bool IsBlock { get; set; } = false;

    public override (Guid ConversationId, Guid UserId) GetId() => (ConversationId, UserId);
}
